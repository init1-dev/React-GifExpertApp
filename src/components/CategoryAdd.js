import React, { useState } from 'react';
import PropTypes from 'prop-types';

const CategoryAdd = ({ categories, setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = ( e ) => {
        setInputValue( e.target.value );
    };

    const handleSubmit = ( e ) => {
        e.preventDefault();

        // This could be done as helper function
        if( inputValue.trim().length > 2 && !undefined ) {
            if(categories.includes(inputValue)) {
                alert('This category already exist');
                setInputValue('');
            } else {
                setCategories( cats => [ inputValue, ...cats ] );
                setInputValue('');
            }
        }
    };

    return (
        <form onSubmit={ handleSubmit }>
            <input
                type="text"
                name="categories"
                id="categories"
                placeholder="Search Categories"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}

export default CategoryAdd;

CategoryAdd.propTypes = {
    categories: PropTypes.array.isRequired,
    setCategories: PropTypes.func.isRequired
}