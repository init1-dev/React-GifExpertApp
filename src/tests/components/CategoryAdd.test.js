import { shallow } from 'enzyme';
import CategoryAdd from '../../components/CategoryAdd';

describe('Pruebas en <CategoryAdd.js />', () => {

    // const setCategories = () => {};

    const setCategories = jest.fn();

    let wrapper;
    
    beforeEach( () => {
        jest.clearAllMocks();

        wrapper = shallow( <CategoryAdd setCategories={ setCategories } /> );
    })

    test('Debe mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe cambiar la caja de texto', () => {
        const value = 'Hola Mundo';

        wrapper.find('input').simulate( 'change', { target: { value } } );
        
        expect( wrapper.find('input').prop('value') ).toBe( value );
    });
    
    test('NO debe postear la información con submit', () => {
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect( setCategories ).not.toHaveBeenCalled();
    });
    
    test('Debe llamar el setCategories y limpiar la caja de texto', () => {
        const value = 'Hola Mundo';

        // 1. Simular el imputChange
        wrapper.find('input').simulate( 'change', { target: { value } } );

        // 2. Simular el submit
        wrapper.find('form').simulate( 'submit', { preventDefault(){} });

        // 3. setCategories se debe de haber llamado
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

        // 4. El valor del input debe ser un string vacío
        expect( wrapper.find('input').prop('value') ).toBe('');
    });
    
});