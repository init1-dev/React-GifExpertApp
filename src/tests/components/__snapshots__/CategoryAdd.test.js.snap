// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CategoryAdd.js /> Debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Search Categories"
    type="text"
    value=""
  />
</form>
`;
